HttpTrigger1/auth.py

```diff
@@ -0,0 +1,80 @@
+import os
+import json
+from urllib.request import urlopen
+from jose import jwt
+
+AUTH0_DOMAIN = os.getenv("AUTH0_DOMAIN")
+API_AUDIENCE = os.getenv("API_AUDIENCE")
+ALGORITHMS = ["RS256"]
+PERMISSION = os.getenv("PERMISSION")
+
+
+class AuthError(Exception):
+    def __init__(self, message, status_code):
+        self.message = message
+        self.status_code = status_code
+
+
+def get_token_auth_header(req):
+    auth_header = req.headers.__http_headers__.get("authorization", None)
+    if auth_header is None:
+        raise AuthError("Authorization header is missing", 401)
+    header_parts = auth_header.split()
+    if header_parts[0].lower() != "bearer":
+        raise AuthError("Invalid Header: must start with Bearer.", 401)
+    if len(header_parts) == 1:
+        raise AuthError("Invalid Header: Missing token", 401)
+    if len(header_parts) > 2:
+        raise AuthError("Invalid Header: Must be bearer token", 401)
+    jwt_token = header_parts[1]
+    return jwt_token
```

Comment:
Avoid using exception handling as a control flow mechanism: in this function, you can return a tuple of `AuthError` and the `token` and then check if there's an error.

```suggestion
def get_token_auth_header(req) -> Tuple[Union[AuthError, None], str]:
    auth_header = req.headers.__http_headers__.get("authorization", '')
    header_parts = auth_header.split()
    
    if auth_header is None:
        return AuthError("Authorization header is missing", 401), ''
    if header_parts[0].lower() != "bearer":
        return AuthError("Invalid Header: must start with Bearer.", 401), ''
    if len(header_parts) == 1:
        return AuthError("Invalid Header: Missing token", 401), ''
    if len(header_parts) > 2:
        return AuthError("Invalid Header: Must be bearer token", 401), ''

    return None, header_parts[1]
```

Reply:
What is the rationale behind avoiding using exception handling?
The suggested solution doesn't seem to add any extra functionality, only extra complexity 🔢